### PR TITLE
Framework: Add a `context.render` wrapper for `ReactDom.render`

### PR DESCRIPTION
**Pull request by Tug**
This PR adds a `context.render` function, a simple wrapper for `ReactDom.render` which injects the `ReduxStore` using `react-redux`'s `Provider`.

This could be another way of fixing issue https://github.com/Automattic/wp-calypso/issues/1943
So let's discuss it!

### CODE REVIEW

**Comment by Tug** [comment]
cc: @drewblaisdell @artpi @aduth @nb

**Comment by drewblaisdell** [comment]
The benefit of this method is that it reduces boilerplate code and could prevent us from missing areas where we forget provide the global state to components that need to access it or have descendants that need to access it if we always use `context.render`. Is that correct?

**Comment by Tug** [comment]
Yes, that's the idea: replace every `ReactDom.render` calls by `context.render`, so issue #1943 won't appear again as we convert components to the new global redux store.

**Comment by artpi** [comment]
Nice idea!
I completely agree this is needed and would be VERY helpful with my further reduxification efforts like in #1916 , so I'm very keen on having this kind of 

However, I dont know if context is a place for it.
It is for sure convenient, but cant we make it a pure function and turn it into some kind of helper/mixin/lib ?
Just asking - I'm ok for having it in context, but debugging it can be a pain, because it does a bit of magic and figuring out why the return of this behaves weird.
(And it will behave wierd because it will return renderes Provider - not something we have pased as an argument)

File: client/boot/index.js [hunk 1]
```diff
@@ -88,6 +88,12 @@ function setUpContext( layout, reduxStore ) {
 
 		context.layout = layout;
 		context.store = reduxStore;
+		context.render = ( reactElement, domContainerNode ) => {
```
Comment (artpi) [inline client/boot/index.js]: Maybe `renderWithProvider`, `reduxProviderRender` or something similar?

**Comment by Tug** [comment]
@artpi I agree that context might not be the best place for this helper function or that `render` is not descriptive enough. The idea was to add a really simple function already accessible from every component.
Ok let's do it as an external module, it will be cleaner and if every `ReactDom.render` is replaced with this helper, its purpose won't be too hard to notice.

About the magic part, I'm not sure it will make debugging harder (we can use a named function if you want) but it can indeed lead to unexpected results in tests or other code which uses the return value of the renderer.

**Comment by artpi** [comment]
@Tug : When I wrapped `contex.layout` in `boot.js` in  Provider, it turned out that some modules were calling `context.layout.setState`, which changed state in Provider, not in Layout and showing/hiding sidebar seized to work.
That's what I mean by sideffect :)

**Comment by gziolo** [comment]
I agree that this helper method shouldn't be a part of the `context`, but definitely it's great way of removing some code boilerplate :) My vote is also for external function.

File: client/me/controller.js [hunk 2]
```diff
@@ -281,7 +281,7 @@ export default {
 		if ( transactionId ) {
 			analytics.pageView.record( basePath + '/receipt', ANALYTICS_PAGE_TITLE + ' > Billing History > Receipt' );
 
-			ReactDom.render(
+			context.render(
```
Comment (gziolo) [inline client/me/controller.js]: we could call it this way:
```js
const reactElement = React.createElement( ViewReceiptModal, { transaction: billingData.getTransaction( transactionId ) } );
renderWithProvider( reactElement, 'tertiary', context.store );
```
I guess we always use `document.getElementById` method, so we could simplify also that part.

File: client/boot/index.js [hunk 3]
```diff
@@ -88,6 +88,12 @@ function setUpContext( layout, reduxStore ) {
...
+			return ReactDom.render(
```
Comment (gziolo) [inline client/boot/index.js]: I'm not sure if we need return value here. We don't use it at the moment.

Comment (artpi) [inline client/boot/index.js]: If we intend to replace ReactDom.render, then we need to kinda return the same I think..

**Comment by Tug** [comment]
I created a new PR following your suggestions https://github.com/Automattic/wp-calypso/pull/1955
Feel free to review it when you have time.

**Pull request closed without merging**